src/asl.ts

```diff
@@ -2837,60 +2792,494 @@ export class ASL {
+              },
+              ResultPath: filterResult,
+              Next: "set",
+            },
+            // filterResult = filterResult.result
+            set: {
+              Type: "Pass",
+              // the original array is initialized with a leading null to simplify the adding of new values to the array string "[null"+",{new item}"+"]"
+              InputPath: `${filterResult}.result[1:]`,
+              ResultPath: filterResult,
+              Next: ASLGraph.DeferNext,
+            },
+          },
+        },
+      },
+      output: { jsonPath: filterResult },
+    };
+  }
+
+  private mapToStateOutput(
+    expr: CallExpr & {
+      expr: PropAccessExpr;
+    }
+  ) {
+    const callbackfn = expr.args[0].expr;
+    if (!isFunctionLike(callbackfn)) {
       throw new SynthError(
-        ErrorCodes.StepFunction_invalid_filter_syntax,
-        `JSONPath's filter expression does not support '${nodeToString(expr)}'`
+        ErrorCodes.Invalid_Input,
+        `the 'callback' argument of map must be a function or arrow expression, found: ${callbackfn?.kindName}`
       );
-    };
+    }
+    return this.evalExpr(
+      expr.expr.expr,
+      (listOutput, { normalizeOutputToJsonPath }) => {
+        // we assume that an array literal or a call would return a variable.
+        if (
+          ASLGraph.isLiteralValue(listOutput) &&
+          !Array.isArray(listOutput.value)
+        ) {
+          throw new SynthError(
+            ErrorCodes.Unexpected_Error,
+            "Expected input to map to be a variable reference or array"
+          );
+        }
+
+        const listPath = normalizeOutputToJsonPath().jsonPath;
+
+        const mapResultVariable = this.newHeapVariable();
+
+        const [valueParameter, indexParameter, arrayParameter] =
+          callbackfn.parameters;
+
+        const uniqueReturnName = this.generatedNames.generateOrGet(expr);
+
+        const bodyStates = ASLGraph.joinSubStates(
+          callbackfn,
+          // run any parameter initializers if they exist
+          valueParameter
+            ? this.evalDecl(valueParameter, {
+                jsonPath: indexParameter
+                  ? `${mapResultVariable}.arr[0].item`
+                  : `${mapResultVariable}.arr[0]`,
+              })
+            : undefined,
+          indexParameter
+            ? this.evalDecl(indexParameter, {
+                jsonPath: `${mapResultVariable}.arr[0].index`,
+              })
+            : undefined,
+          arrayParameter
+            ? this.evalDecl(arrayParameter, {
+                jsonPath: listPath,
+              })
+            : undefined,
+          this.evalStmt(callbackfn.body, {
+            End: undefined,
+            // write over the input value to avoid creating a new heap value
+            ResultPath: `${mapResultVariable}.arr[0]`,
+            // unique return name
+            Next: uniqueReturnName,
+          })
```

Comment:
Is this redundant? I've seen this code elsewhere, or at least similar code.

Reply:
Uhh, yeah, map, filter, and forEach are similar. I decided to keep them separate to improve the readability, I'll see where I can merge them. 

Filter:

1. `init`
   1. copy array to tail
   2. initialize result string - map and filter, foreach doesn't have a result
2. `check` for an empty array - if so -> end
3. Setup `parameters`:
   1. item - evalDecl (catches binds)
   2. index - zip indices, evalDecl (bind isn't needed for a number, but just in case)
   3. array - evalDecl (catches bind)
4. run `body`
   3. return to heap value - map and filter
   4. return null - forEach
5. `tail`
   6. always - add body result to result string - map
   7. based on the predicate - add input item to the result string - filter
   8. tail array
6. go to `check`
7. end
   11. clear scratch space - forEach
   12. format result array - map and filter
   13. return result array - map and filter

Reply:
Alright, created a helper that abstracts most of the duplication away. Customization:
1. Whether to create a new variable for the callback output or write over the original item (in temp array)
2. how to handler the result of the callback and tail
3. what to do at the end (format array string vs nothing).